nautobot/extras/models/groups.py

```diff
@@ -431,3 +446,199 @@ def clean(self):
+
+            operator = membership.operator
+
+            logger.debug("%s -> %s -> %s", group, group.filter, operator)
+
+            # The previous iteration that used set operations at the DB layer
+            # but limited by not being able to do further filtering.
+            """
+            fs = self.filterset_class(group.filter, self.get_queryset())
+            if operator == "union":
+                qs = qs.union(fs.qs)
+            elif operator == "difference":
+                qs = qs.difference(fs.qs)
+            elif operator == "intersection":
+                qs = qs.intersection(fs.qs)
+            """
+            next_set = self.generate_query_for_group(group)
+            # print(next_set)
+            if operator == "union":
+                big_q |= next_set
+                qs = qs | base_qs.filter(next_set)
+            elif operator == "difference":
+                big_q &= ~next_set
+                qs = qs.exclude(next_set)
+            elif operator == "intersection":
+                big_q &= next_set
+                qs = qs.filter(next_set)
+            # print(big_q)
+
+        # return big_q
+        return qs
+
+    def get_group_queryset(self):
+        return self.process_group_filters(group=self)
+
+    def get_descendants(self, group=None, descendants=None):
+        """Return the children of all child groups."""
+        if group is None:
+            group = self
+        if descendants is None:
+            descendants = []
+
+        for child_group in group.children.all():
+            print(f"Processing group {child_group}...")
+            # descendants.append(child_group.pk)
+            descendants.append(child_group)
+            if child_group.children.exists():
+                self.get_descendants(child_group, descendants)
+
+        # TODO(jathan): Return a QuerySet with default ordering? Or a list that
+        # is explicitly ordered by distance from object.
+        return descendants
+        # return DynamicGroup.objects.filter(pk__in=descendants)
+
+    def get_ancestors(self, group=None, ancestors=None):
+        """Return the parents of all parent groups."""
+        if group is None:
+            group = self
+        if ancestors is None:
+            ancestors = []
+
+        for parent_group in group.parents.all():
+            print(f"Processing group {parent_group}...")
+            # ancestors.append(parent_group.pk)
+            ancestors.append(parent_group)
+            if parent_group.parents.exists():
+                self.get_ancestors(parent_group, ancestors)
+
+        # TODO(jathan): Return a QuerySet with default ordering? Or a list that
+        # is explicitly ordered by distance from object.
+        return ancestors
+        # return DynamicGroup.objects.filter(pk__in=ancestors)
+
+    def get_siblings(self, parent_group=None):
+        """Return groups that share the same parents."""
+        if parent_group is None:
+            parent_group = self
```

Comment:
Should this be `self.parents.all()` or something other than `self`? It doesn't seem right to me as written.

Reply:
It's implying that if you call it without arguments it starts with itself. You're right that this is not correct as implemented however. It's a bit naive.